src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
             return start < TextWindow.Position;
         }
 
+        // Allows underscores in integers, except at beginning and end
+        private void ScanNumericLiteralSingleInteger(StringBuilder builder, ref bool underscoreInWrongPlace, ref bool usedUnderscore, bool isHex, bool isBinary)
+        {
+            if (TextWindow.PeekChar() == '_')
+            {
+                underscoreInWrongPlace = true;
+            }
+
+            char ch;
+            var lastCharWasUnderscore = false;
+            while (true)
+            {
+                ch = TextWindow.PeekChar();
+                if (ch == '_')
+                {
+                    usedUnderscore = true;
+                    lastCharWasUnderscore = true;
+                }
+                else
+                {
+                    if (isHex)
+                    {
+                        if (!SyntaxFacts.IsHexDigit(ch))
+                        {
+                            break;
+                        }
+                    }
+                    else if (isBinary)
+                    {
+                        if (!SyntaxFacts.IsBinaryDigit(ch))
+                        {
+                            break;
+                        }
+                    }
+                    else
+                    {
+                        if (!SyntaxFacts.IsDecDigit(ch))
+                        {
+                            break;
+                        }
+                    }
+                    _builder.Append(ch);
+                    lastCharWasUnderscore = false;
+                }
+                TextWindow.AdvanceChar();
+            }
```

Comment:
Just tried it, I don't think the new version is equivalent (test cases failing). Example:

Say we're parsing a hex, so isHex == true, and the current character is 'a'. `(isHex && !SyntaxFacts.IsHexDigit(ch))` is false, so it moves to the next or. `isBinary` is false, so skips that too. `SyntaxFacts.IsDecDigit('a')` is false, so the `!` makes it true, so the entire expression is true and we break out of the loop incorrectly.

I do like the shortened version, but I'm not sure how to fix it to not "accidentally jump into other digit's conditions" without getting very verbose/obfuscated.